KEP-4017: promote Pod Index Label for StatefulSets and Indexed Jobs to GA

alaypatel07 opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: promote KEP-4017 to GA

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4017

<!-- other comments or additional information -->
- Other comments:

kannon92 commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -411,7 +411,8 @@ existing label which other things may depend on, for example).
 #### GA
-Fix any potentially reported bugs.
+- Fix report doc bug: https://github.com/kubernetes/kubernetes/pull/125502/
+- Add IndexJob pod label (same as existing annotation)
```
I see this code:

```
func addCompletionIndexLabel(template *v1.PodTemplateSpec, index int) {
	if template.Labels == nil {
		template.Labels = make(map[string]string, 1)
	}
	// For consistency, we use the annotation batch.kubernetes.io/job-completion-index for the corresponding label as well.
	template.Labels[batch.JobCompletionIndexAnnotation] = strconv.Itoa(index)
}
```

I think IndexJob pod label has already been added in beta.

alaypatel07 replied:
thanks for checking, removed from the list

kannon92 commented:
/ok-to-test

Only thing I wonder is if we have to cherry-pick the release notes.

The website seems to have the right labels.

kannon92 commented:
/cc @soltysh @mimowo @atiratree

kannon92 commented:
LGTM to be but I'll leave the labels for a sig-apps reviewer/approver.

mimowo commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -411,7 +411,7 @@ existing label which other things may depend on, for example).
 
 #### GA
-Fix any potentially reported bugs.
+- Fix report doc bug: https://github.com/kubernetes/kubernetes/pull/125502/
```
I would not block it on it. We probably don't need it at all at this point as suggested there: https://github.com/kubernetes/kubernetes/pull/125502#issuecomment-2367450201.

soltysh replied:
It's a minor fix, I honestly wouldn't even mention it here :wink: but I won't object either.

alaypatel07 replied:
removed this to make it simpler

mimowo commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -411,7 +411,7 @@ existing label which other things may depend on, for example).
 - Docs are clear about what happens if two pods get the same value (it is set by workload controllers, nothing in the API system will prevent collisions from happening).
 
 #### GA
```
Please add a point about 
Locking the `PodIndexLabel` feature-gate, and modifying the code to ignore it

alaypatel07 replied:
Added, PTAL

kannon92 replied:
I don't see a change here.

alaypatel07 replied:
missed pushing it, should be there now

mimowo replied:
thanks, lgtm

mimowo reviewed (commented):
Please also:
- update the checkboxes in "Release Signoff Checklist"
- update "Implementation history"

wojtek-t commented on keps/prod-readiness/sig-apps/4017.yaml:
```diff
@@ -1,3 +1,5 @@
 kep-number: 4017
 beta:
+  approver: "@wojtek-t"
+stable:
```
Can't comment on unchanged lines, so adding the questions here:

> It will be tested manually prior to beta launch.

Has this happened?
Please update the response to describe the test together with the findings.

kannon92 replied:
Cc @danielvegamyhre

alaypatel07 replied:
@wojtek-t I tracked back the feature in beta to this PR https://github.com/kubernetes/kubernetes/pull/119232 but I do not find any comments on the manual testing done.

As a way to move forward, once the feature gate it locked, would it be okay to document the results as upgrade from beta -> GA of this feature?

kannon92 replied:
@wojtek-t how does one test upgrades with local environments?

wojtek-t replied:
I think it's already a bit too late - if we push that as GA criteria, I'm afraid it will just not happen at all.

Regarding testing with local environments, there are many examples of that across different KEPs. A random example with kind is here:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1669-proxy-terminating-endpoints#were-upgrade-and-rollback-tested-was-the-upgrade-downgrade-upgrade-path-tested

It shouldn't really take more than an hour or so to do some manual test and validate it.

alaypatel07 replied:
@wojtek-t from your provided example, I have updated the `###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?` section, PTAL

soltysh commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -238,8 +238,8 @@ At a high level, the proposal is to modify the StatefulSet and Job controllers t
-StatefulSet pod label: `statefulset.kubernetes.io/pod-index`
-Indexed Job pod label: `batch.kubernetes.io/job-completion-index` (same as existing annotation)
+- StatefulSet pod label: `apps.kubernetes.io/pod-index`
+- IndexedJob pod label: `batch.kubernetes.io/job-completion-index` (same as existing annotation)
```
:+1:

soltysh commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -238,8 +238,8 @@ At a high level, the proposal is to modify the StatefulSet and Job controllers t
 as a pod label at pod creation time (for jobs, this would only apply to jobs in
 Indexed completion mode). The details of this are outlined in the Design Details section below.
 
-StatefulSet pod label: `statefulset.kubernetes.io/pod-index`
```
Nit, but would be good to make sure we have this checked out, in the `## Release Signoff Checklist` section make sure to check all the appropriate boxes.

soltysh commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -238,8 +238,8 @@ At a high level, the proposal is to modify the StatefulSet and Job controllers t
+- StatefulSet pod label: `apps.kubernetes.io/pod-index`
+- IndexedJob pod label: `batch.kubernetes.io/job-completion-index` (same as existing annotation)
 
 ### User Stories (Optional)
```
Nit, in the 2nd story, there's a typo, we have:
```
Instead of creating a service which matche an entire Job
```
but should be
```
Instead of creating a service which matches an entire Job
```

alaypatel07 replied:
updated.

soltysh commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -311,7 +311,7 @@ change are understandable. This may include API specs (though not always
+The StatefulSet controller will only need a minor update to the [newStatefulSetPod](https://github.com/kubernetes/kubernetes/blob/fb5e9ef3b2a6f2136d54868187431c345e59f55f/pkg/controller/statefulset/stateful_set_utils.go#L458) function, to set the pod ordinal as the label `apps.kubernetes.io/pod-index`. This call is downstream from the [newVersionedStatefulSetPod](https://github.com/kubernetes/kub ...
 the StatefulSet pods before creating them as necessary in [CreateStatefulPod](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controller/statefulset/stateful_set_control.go#L433).
 
 Similarly, the Job controller would need to add the completion index as a label [here](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controller/job/job_controller.go#L1480)
```
Below:
- units - can you update the coverage based on current data
- integration and/or e2e - can you link the results from the tests that were added covering this new functionality

alaypatel07 replied:
@soltysh I found the unit test covering this, I couldnt find the e2e/integration test.

@kannon92 or @soltysh can you help me find where to look for integration/e2e tests for statefulset controller?

kannon92 replied:
https://github.com/kubernetes/enhancements/pull/4865#issuecomment-2398046329

Integration tests are usually under test/integration and e2e tests are under test/e2e.

soltysh replied:
Function in integration checking that value is here: https://github.com/kubernetes/kubernetes/blob/d9c46d8ecb1ede9be30545c9803e17682fcc4b50/test/integration/job/job_test.go#L4362C6-L4371 so a test exists as well. 
E2E - https://github.com/kubernetes/kubernetes/blob/d9c46d8ecb1ede9be30545c9803e17682fcc4b50/test/e2e/apps/job.go#L435-L467

soltysh commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -411,7 +411,7 @@ existing label which other things may depend on, for example).
+- Fix report doc bug: https://github.com/kubernetes/kubernetes/pull/125502/
 
 <!--
 **Note:** Generally we also wait at least two releases between beta and
```
Below, in `### Version Skew Strategy`, we have:
```
and the node componets were at version N-1 where this feature does not exist
```
but should be:
```
and the node components were at version N-1 where this feature does not exist
```

alaypatel07 replied:
updated, thanks for the catch

soltysh commented on keps/sig-apps/4017-pod-index-label/kep.yaml:
```diff
@@ -12,18 +12,18 @@ approvers:
-  stable:
+  stable: "v1.32"
 
 feature-gates:
```
One more nit, can you copy the metrics from the SLI question from PRR at the bottom of this document, in the `metrics` section as well.

alaypatel07 replied:
done, PTAL

soltysh reviewed (commented):
The biggest missing are tests (unit, integration/e2e and upgrade->downgrade). Aside from that a few minor nits to address.

alaypatel07 commented:
> Please also:

> update the checkboxes in "Release Signoff Checklist"
> update "Implementation history"
> 

@mimowo I have updated the Release Signoff, pending the e2e tests. @kannon92 mentioned that integration and/or e2e tests were missed for this feature so I will be adding them as part of moving this to GA.

I have updated the Implementation history, PTAL

mimowo commented:
sgtm, thanks!

mimowo reviewed (commented):
LGTM

wojtek-t commented:
/approve PRR

Still requires SIG approval. @soltysh

soltysh commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -333,6 +333,8 @@ when drafting this test plan.
 existing tests to make this code solid enough prior to committing the changes necessary
 to implement this enhancement.
+[X] unit tests: https://github.com/kubernetes/kubernetes/blob/585971431b4c8ea290e8c754d5806b51630a1938/pkg/controller/statefulset/stateful_set_control_test.go#L246-L250
+[X] integration and/or e2e: TODO
```
Have a look at the [template](https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md#test-plan) and update it accordingly.

soltysh commented on keps/sig-apps/4017-pod-index-label/README.md:
```diff
@@ -882,6 +947,8 @@ Major milestones might include:
 -->
 - 2023-05-17: KEP published
+- 2023-07-14: Feature merged with feature gate in beta
+- v1.32.0: Feature graduated to GA
```
Date, not a release :wink:

soltysh reviewed (commented):
@alaypatel07 update test section, with appropriate links from the template. Once you update ping me for tag on slack

soltysh reviewed (approved):
/lgtm
/approve
